morethuente.go

```diff
@@ -0,0 +1,384 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package optimize
+
+import "math"
+
+// MoreThuente is a Linesearcher that finds steps that satisfy both the
+// sufficient decrease and curvature conditions (the strong Wolfe conditions).
+//
+// References:
+//  - More, J.J. and D.J. Thuente: Line Search Algorithms with Guaranteed Sufficient
+//    Decrease. ACM Transactions on Mathematical Software 20(3) (1994), 286-307
+type MoreThuente struct {
+	// DecreaseFactor is the constant factor in the sufficient decrease
+	// (Armijo) condition.
+	// It must be in the interval [0, 1). The default value is 0.
+	DecreaseFactor float64
+	// CurvatureFactor is the constant factor in the Wolfe conditions. Smaller
+	// values result in a more exact line search.
+	// A set value must be in the interval (0, 1). If it is zero, it will be
+	// defaulted to 0.9.
+	CurvatureFactor float64
+	// StepTolerance sets the minimum acceptable width for the linesearch
+	// interval. If the relative interval length is less than this value,
+	// ErrLinesearcherFailure is returned.
+	// It must be non-negative. If it is zero, it will be defaulted to 1e-10.
+	StepTolerance float64
+
+	// MinimumStep is the minimum step that the linesearcher will take.
+	// It must be non-negative and less than MaximumStep. Defaults to no
+	// minimum (a value of 0).
+	MinimumStep float64
+	// MaximumStep is the maximum step that the linesearcher will take.
+	// It must be greater than MinimumStep. If it is zero, it will be defaulted
+	// to 1e20.
+	MaximumStep float64
+
+	bracketed bool    // Indicates if a minimum has been bracketed.
+	fInit     float64 // Function value at step = 0.
+	gInit     float64 // Derivative value at step = 0.
+
+	// When stage is 1, the algorithm updates the interval given by x and y
+	// so that it contains a minimizer of the modified function
+	//  psi(step) = f(step) - f(0) - DecreaseFactor * step * f'(0).
+	// When stage is 2, the interval is updated so that it contains a minimizer
+	// of f.
+	stage int
```

Comment:
In fact it has three values because I (ab)use the zero value of `stage` to check that `Init` has been called. If you don't like it or if such stringent checking is deemed too pedantic, I can make it a bool.